lib/modules/datasource/hackage/index.ts

```diff
@@ -0,0 +1,61 @@
+import is from '@sindresorhus/is';
+import { joinUrlParts } from '../../../util/url';
+import * as pvpVersioning from '../../versioning/pvp';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { HackagePackageMetadata } from './schema';
+
+export class HackageDatasource extends Datasource {
+  static readonly id = 'hackage';
+
+  constructor() {
+    super(HackageDatasource.id);
+  }
+
+  override readonly defaultVersioning = pvpVersioning.id;
+  override readonly customRegistrySupport = false;
+  override readonly defaultRegistryUrls = ['https://hackage.haskell.org/'];
+
+  async getReleases(config: GetReleasesConfig): Promise<ReleaseResult | null> {
+    if (!is.nonEmptyString(config.registryUrl)) {
+      return null;
+    }
+    const massagedPackageName = encodeURIComponent(config.packageName);
+    const url = joinUrlParts(
+      config.registryUrl,
+      'package',
+      `${massagedPackageName}.json`,
+    );
+    const res = await this.http.getJson(url, HackagePackageMetadata);
+    const keys = Object.keys(res.body);
+    return {
+      releases: keys.map((version) =>
+        versionToRelease(version, config.packageName, config.registryUrl!),
+      ),
+    };
+  }
+}
+
+export function versionToRelease(
+  version: string,
+  packageName: string,
+  registryUrl: string,
+): Release {
+  return {
+    version,
+    releaseTimestamp: null,
+    isStable: true,
+    changelogUrl: joinUrlParts(
+      registryUrl,
+      'package',
+      `${packageName}-${version}`,
+      'changelog',
+    ),
+    sourceUrl: joinUrlParts(
+      registryUrl,
+      'package',
+      `${packageName}-${version}`,
+      'src',
+    ),
```

Comment:
Oh, so that's what you meant by source. One issue is that the git repo is specified for each version, such that the project could move hoster over time. So e.g. v1 would be at Github, but v2 would be in Bitbucket. I could add the `source_repository` field to the JSON page returned for each version, [example JSON, note that it is not present](https://hackage.haskell.org/package/lens-5.2.json). The information is already exposed through HTML in the sidebar at e.g. https://hackage.haskell.org/package/lens-5.2. To have it available in Renovate we'd need to do an additional request, to fetch the metadata for *each* version (in the previous example this was done by making the URL terminate with `-5.2.json` instead of just `.json`). Or we could discuss making a whole separate endpoint just for Renovate where we'd return all the necessary information for each version, with just one request. But I feel that that would be best done in a follow-up PR?

Reply:
In that drop in that case for now. 

In that case responses for versions can be cached for a long time.
This should be considered to be implemented as the additional metadata unlocks some features in Renovate like `minimumReleaseAge`.

From Renovates perspective it would ideal if we would get all this data at once.